### PR TITLE
[SPARK-32629][SQL] Track metrics of BitSet/OpenHashSet in full outer SHJ

### PR DESCRIPTION
**Pull request by c21**
### What changes were proposed in this pull request?

This is followup from https://github.com/apache/spark/pull/29342, where to do two things:
* Per https://github.com/apache/spark/pull/29342#discussion_r470153323, change from java `HashSet` to spark in-house `OpenHashSet` to track matched rows for non-unique join keys. I checked `OpenHashSet` implementation which is built from a key index (`OpenHashSet._bitset` as `BitSet`) and key array (`OpenHashSet._data` as `Array`). Java `HashSet` is built from `HashMap`, which stores value in `Node` linked list and by theory should have taken more memory than `OpenHashSet`. Reran the same benchmark query used in https://github.com/apache/spark/pull/29342, and verified the query has similar performance here between `HashSet` and `OpenHashSet`.
* Track metrics of the extra data structure `BitSet`/`OpenHashSet` for full outer SHJ. This depends on above thing, because there seems no easy way to get java `HashSet` memory size.

### Why are the changes needed?

To better surface the memory usage for full outer SHJ more accurately.
This can help users/developers to debug/improve full outer SHJ.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Added unite test in `SQLMetricsSuite.scala` .

### CODE REVIEW

**Comment by c21** [comment]
cc @maropu and @cloud-fan if you guys have time to take a look, thanks.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsTestUtils.scala [hunk 1]
```diff
@@ -254,6 +254,24 @@ trait SQLMetricsTestUtils extends SQLTestUtils {
...
     }
   }
+
+  /**
+   * Verify if the metrics in `SparkPlan` operator are same as expected metrics.
+   *
+   * @param plan `SparkPlan` operator to check metrics
+   * @param expectedMetrics the expected metrics. The format is `metric name -> metric value`.
+   */
+  protected def testMetricsInSparkPlanOperator(
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsTestUtils.scala:264]: We need to put this func here instead of `SQLMetricsSuite`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/ShuffledHashJoinExec.scala [hunk 2]
```diff
@@ -150,9 +148,9 @@ case class ShuffledHashJoinExec(
       streamNullJoinRowWithBuild: => InternalRow => JoinedRow,
       buildNullRow: GenericInternalRow,
       streamNullRow: GenericInternalRow): Iterator[InternalRow] = {
-    // TODO(SPARK-32629):record metrics of extra BitSet/HashSet
-    // in full outer shuffled hash join
     val matchedKeys = new BitSet(hashedRelation.maxNumKeysIndex)
+    val buildDataSize = longMetric("buildDataSize")
+    buildDataSize += matchedKeys.capacity / 8
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/ShuffledHashJoinExec.scala:153]: nit: `longMetric("buildDataSize") += matchedKeys.capacity / 8`?

**Comment by maropu** [comment]
Looks okay if the tests pass. also cc: @agrawaldevesh

File: sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsTestUtils.scala (hunk 1 above)
Comment (c21) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsTestUtils.scala:264]: @maropu - I am following the convention like other method e.g. `testSparkPlanMetrics`.

**Comment by maropu** [comment]
Basically, we don't need the `[FOLLOWUP]` tag for a new ticket, I think.

**Comment by c21** [comment]
> Basically, we don't need the [FOLLOWUP] tag for a new ticket, I think.

@maropu - cool, thanks.

**Comment by agrawaldevesh** [comment]
Wow, cool. So any improvement in either memory usage, GC, or CPU time by switching to open hashset ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/ShuffledHashJoinExec.scala [hunk 3]
```diff
@@ -136,10 +134,10 @@ case class ShuffledHashJoinExec(
    * Full outer shuffled hash join with unique join keys:
    * 1. Process rows from stream side by looking up hash relation.
    *    Mark the matched rows from build side be looked up.
-   *    A `BitSet` is used to track matched rows with key index.
+   *    A [[BitSet]] is used to track matched rows with key index.
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/ShuffledHashJoinExec.scala:137]: nit: just say `A bit set is ...` to be generic.

Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/ShuffledHashJoinExec.scala:137]: nit: just say `A bit set is ...` to be generic.

Comment (c21) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/ShuffledHashJoinExec.scala:137]: @cloud-fan - sure, updated.

Comment (c21) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/ShuffledHashJoinExec.scala:137]: @cloud-fan - sure, updated.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/ShuffledHashJoinExec.scala (hunk 2 above)
Comment (c21) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/ShuffledHashJoinExec.scala:153]: @maropu - sure, updated.

**Comment by c21** [comment]
> So any improvement in either memory usage, GC, or CPU time by switching to open hashset ?

@agrawaldevesh - here is report for one production query internally which is doing a FULL OUTER JOIN for one large table, and one small table. Couldn't share the exact query text and data but the query shape is:

```sql
INSERT OVERWRITE TABLE output_table
PARTITION (...)
SELECT ...
FROM large_table a
FULL OUTER JOIN small_table b
ON a.col_x = f.col_y
```

input metrics:
large_table (ORC format): uncompressed data input size: 54TB
small_table (ORC format): uncompressed data input size: 85GB

execution metrics:

Number of tasks per stage:
stage 0: 40547 (read large table)
stage 1: 158 (read small table)
stage 2: 5063 (SHJ and insert to output table)

Total shuffle bytes across executors: 15.1TB.

Query type | Aggregated executors CPU time (ms) | Aggregated executors GC time (ms)
------------ | ------------- | -------------
use java `HashSet` | 3.48 B | 124.0 M
use spark `OpenHashSet` | 3.22 B | 91.2 M 

TLDR is switching to `OpenHashSet`, we are seeing 7% CPU reduction and 27% GC time reduction.

**Comment by c21** [comment]
Address all comments and PR is ready for review again, thanks.

**Review by cloud-fan — APPROVED** [review]

**Review by maropu — APPROVED** [review]

**Comment by maropu** [comment]
Thanks! Merged to master.

**Comment by c21** [comment]
Thanks @maropu , @cloud-fan and @agrawaldevesh for discussion and review!
